### PR TITLE
[ENG-2129] GET Api Endpoints for Registration Moderation

### PR DESCRIPTION
**Pull request by Johnetordoff**
## Purpose

This allows all relevant registries moderation info to be accessible via the API.

## Changes

- add endpoint for registration provider permissions
- add endpoint for requests (AKA withdraw requests)
- add endpoint actions (AKA logs for submissions transitions)
- give special permissions for moderators to see submitted registrations

## QA Notes

  - Does this change require a data migration? If so, what data will we migrate?
Nope
  - What is the level of risk?
Low
    - Any permissions code touched?
Yes
    - Is this an additive or subtractive change, other?
additive
  - How can QA verify? (Through UI, API, AdminApp or AdminAdminApp?)
API only
    - If verifying through API, what's the new version? Please include the endpoints in PR notes or Dev docs.

  - How will this impact performance?
Not
-->

## Documentation

API docs forthcoming.
| url | class | doc |
| --- | --- | --- |
| `/<version>/providers/registrations/<provider_id>/actions/` | RegistrationProviderActionList | These should be used like logs to track different transitions, initiators and project histories  |
| `/<version>/providers/registrations/<provider_id>/registrations/` | RegistrationProviderRegistrationList | This is the list of registrations associated with a provider, it should be filterable by state.  |
| `/<version>/providers/registrations/<provider_id>/requests/` | RegistrationProviderRequestList	|  These is a list of all the requests made against registrations for a registration provider, current it should only show withdrawal requests.  |

Sample Json :

**RegistrationProviderActionList**

```json
{
   "data":[
      {
         "id":"5f60d0a863d3d300639995e9",
         "type":"node-request-actions",
         "attributes":{
            "trigger":"submit",
            "comment":"",
            "from_state":"initial",
            "to_state":"pending",
            "date_created":"2020-09-15T14:33:12.220483Z",
            "date_modified":"2020-09-15T14:33:12.220539Z",
            "auto":false,
            "visible":true
         },
         "relationships":{
            "creator":{
               "links":{
                  "related":{
                     "href":"http://localhost:8000/v2/users/8zpmd/",
                     "meta":{
                        
                     }
                  }
               },
               "data":{
                  "id":"8zpmd",
                  "type":"users"
               }
            },
            "target":{
               "links":{
                  "related":{
                     "href":"http://localhost:8000/v2/requests/5f60d09d63d3d300639995df/",
                     "meta":{
                        
                     }
                  }
               },
               "data":{
                  "id":"5f60d09d63d3d300639995df",
                  "type":"requests"
               }
            }
         },
         "embeds":{
            "creator":{
               "data":{
                  "id":"8zpmd",
                  "type":"users",
                  "attributes":{
                     "full_name":"Freddie Mercury1",
                     "given_name":"Freddie",
                     "middle_names":"",
                     "family_name":"Mercury1",
                     "suffix":"",
                     "date_registered":"2020-09-15T14:32:59.982483",
                     "active":true,
                     "timezone":"Etc/UTC",
                     "locale":"en_US",
                     "social":{
                        
                     },
                     "employment":[
                        
                     ],
                     "education":[
                        
                     ]
                  },
                  "relationships":{
                     "nodes":{
                        "links":{
                           "related":{
                              "href":"http://localhost:80/v2/users/8zpmd/nodes/",
                              "meta":{
                                 
                              }
                           }
                        }
                     },
                     "groups":{
                        "links":{
                           "related":{
                              "href":"http://localhost:80/v2/users/8zpmd/groups/",
                              "meta":{
                                 
                              }
                           }
                        }
                     },
                     "quickfiles":{
                        "links":{
                           "related":{
                              "href":"http://localhost:80/v2/users/8zpmd/quickfiles/",
                              "meta":{
                                 
                              }
                           },
                           "upload":{
                              "href":"http://localhost:7777/v1/resources/6jn7cst8um/providers/osfstorage/",
                              "meta":{
                                 
                              }
                           },
                           "download":{
                              "href":"http://localhost:7777/v1/resources/6jn7cst8um/providers/osfstorage/?zip=",
                              "meta":{
                                 
                              }
                           }
                        }
                     },
                     "registrations":{
                        "links":{
                           "related":{
                              "href":"http://localhost:80/v2/users/8zpmd/registrations/",
                              "meta":{
                                 
                              }
                           }
                        }
                     },
                     "institutions":{
                        "links":{
                           "related":{
                              "href":"http://localhost:80/v2/users/8zpmd/institutions/",
                              "meta":{
                                 
                              }
                           },
                           "self":{
                              "href":"http://localhost:80/v2/users/8zpmd/relationships/institutions/",
                              "meta":{
                                 
                              }
                           }
                        }
                     },
                     "preprints":{
                        "links":{
                           "related":{
                              "href":"http://localhost:80/v2/users/8zpmd/preprints/",
                              "meta":{
                                 
                              }
                           }
                        }
                     }
                  },
                  "links":{
                     "html":"http://localhost:5000/8zpmd/",
                     "profile_image":"https://secure.gravatar.com/avatar/c406b0a57370bdf201177ff57dba629f?d=identicon",
                     "self":"http://localhost:8000/v2/users/8zpmd/"
                  }
               }
            }
         },
         "links":{
            "self":"http://localhost:8000/v2/actions/5f60d0a863d3d300639995e9/"
         }
      }
   ],
   "links":{
      "first":"None",
      "last":"None",
      "prev":"None",
      "next":"None",
      "meta":{
         "total":1,
         "per_page":10
      }
   },
   "meta":{
      "version":"2.0"
   }
}
```

**RegistrationProviderRegistrationList**

```json
{
   "data":[
      {
         "attributes":{
            "access_requests_enabled":false,
            "archiving":true,
            "article_doi":null,
            "category":"project",
            "collection":false,
            "current_user_can_comment":false,
            "current_user_is_contributor":false,
            "current_user_is_contributor_or_group_member":false,
            "current_user_permissions":[
               "read"
            ],
            "custom_citation":"",
            "date_created":"2020-10-06T12:32:27.795842",
            "date_modified":"2020-10-06T12:32:26.479857",
            "date_registered":"2020-10-06T12:32:27.786988",
            "date_withdrawn":null,
            "description":"Although relationship its beat yard create trial.",
            "embargo_end_date":null,
            "embargoed":false,
            "fork":false,
            "machine_state":"initial",
            "node_license":null,
            "pending_embargo_approval":false,
            "pending_embargo_termination_approval":false,
            "pending_registration_approval":false,
            "pending_withdrawal":false,
            "preprint":false,
            "public":false,
            "registered_meta":{
               
            },
            "registration":true,
            "registration_responses":{
               "summary":""
            },
            "registration_supplement":"Open-Ended Registration",
            "subjects":[
               
            ],
            "tags":[
               
            ],
            "title":"Innovative stable artificial intelligence",
            "wiki_enabled":true,
            "withdrawal_justification":null,
            "withdrawn":false
         },
         "id":"6a9k8",
         "links":{
            "html":"http://localhost:5000/6a9k8/",
            "self":"http://localhost:8000/v2/registrations/6a9k8/"
         },
         "relationships":{
            "affiliated_institutions":{
               "links":{
                  "related":{
                     "href":"http://localhost:8000/v2/registrations/6a9k8/institutions/",
                     "meta":{
                        
                     }
                  },
                  "self":{
                     "href":"http://localhost:8000/v2/registrations/6a9k8/relationships/institutions/",
                     "meta":{
                        
                     }
                  }
               }
            },
            "bibliographic_contributors":{
               "links":{
                  "related":{
                     "href":"http://localhost:8000/v2/registrations/6a9k8/bibliographic_contributors/",
                     "meta":{
                        
                     }
                  }
               }
            },
            "children":{
               "links":{
                  "related":{
                     "href":"http://localhost:8000/v2/registrations/6a9k8/children/",
                     "meta":{
                        
                     }
                  }
               }
            },
            "citation":{
               "data":{
                  "id":"6a9k8",
                  "type":"registrations"
               },
               "links":{
                  "related":{
                     "href":"http://localhost:8000/v2/registrations/6a9k8/citation/",
                     "meta":{
                        
                     }
                  }
               }
            },
            "comments":{
               "links":{
                  "related":{
                     "href":"http://localhost:8000/v2/registrations/6a9k8/comments/?filter%5Btarget%5D=6a9k8",
                     "meta":{
                        
                     }
                  }
               }
            },
            "contributors":{
               "links":{
                  "related":{
                     "href":"http://localhost:8000/v2/registrations/6a9k8/contributors/",
                     "meta":{
                        
                     }
                  }
               }
            },
            "files":{
               "links":{
                  "related":{
                     "href":"http://localhost:8000/v2/registrations/6a9k8/files/",
                     "meta":{
                        
                     }
                  }
               }
            },
            "forks":{
               "links":{
                  "related":{
                     "href":"http://localhost:8000/v2/registrations/6a9k8/forks/",
                     "meta":{
                        
                     }
                  }
               }
            },
            "identifiers":{
               "links":{
                  "related":{
                     "href":"http://localhost:8000/v2/registrations/6a9k8/identifiers/",
                     "meta":{
                        
                     }
                  }
               }
            },
            "implicit_contributors":{
               "links":{
                  "related":{
                     "href":"http://localhost:8000/v2/registrations/6a9k8/implicit_contributors/",
                     "meta":{
                        
                     }
                  }
               }
            },
            "linked_by_nodes":{
               "links":{
                  "related":{
                     "href":"http://localhost:8000/v2/registrations/6a9k8/linked_by_nodes/",
                     "meta":{
                        
                     }
                  }
               }
            },
            "linked_by_registrations":{
               "links":{
                  "related":{
                     "href":"http://localhost:8000/v2/registrations/6a9k8/linked_by_registrations/",
                     "meta":{
                        
                     }
                  }
               }
            },
            "linked_nodes":{
               "links":{
                  "related":{
                     "href":"http://localhost:8000/v2/registrations/6a9k8/linked_nodes/",
                     "meta":{
                        
                     }
                  },
                  "self":{
                     "href":"http://localhost:8000/v2/registrations/6a9k8/relationships/linked_nodes/",
                     "meta":{
                        
                     }
                  }
               }
            },
            "linked_registrations":{
               "links":{
                  "related":{
                     "href":"http://localhost:8000/v2/registrations/6a9k8/linked_registrations/",
                     "meta":{
                        
                     }
                  },
                  "self":{
                     "href":"http://localhost:8000/v2/registrations/6a9k8/relationships/linked_registrations/",
                     "meta":{
                        
                     }
                  }
               }
            },
            "logs":{
               "links":{
                  "related":{
                     "href":"http://localhost:8000/v2/registrations/6a9k8/logs/",
                     "meta":{
                        
                     }
                  }
               }
            },
            "node_links":{
               "links":{
                  "related":{
                     "href":"http://localhost:8000/v2/registrations/6a9k8/node_links/",
                     "meta":{
                        
                     }
                  }
               }
            },
            "provider":{
               "links":{
                  "related":{
                     "href":"http://localhost:8000/v2/providers/registrations/5f7c63d8d99cee00b05f9caf/",
                     "meta":{
                        
                     }
                  }
               }
            },
            "region":{
               "data":{
                  "id":"us",
                  "type":"regions"
               },
               "links":{
                  "related":{
                     "href":"http://localhost:8000/v2/regions/us/",
                     "meta":{
                        
                     }
                  }
               }
            },
            "registered_by":{
               "data":{
                  "id":"uqdxe",
                  "type":"users"
               },
               "links":{
                  "related":{
                     "href":"http://localhost:8000/v2/users/uqdxe/",
                     "meta":{
                        
                     }
                  }
               }
            },
            "registered_from":{
               "data":{
                  "id":"pvx9d",
                  "type":"nodes"
               },
               "links":{
                  "related":{
                     "href":"http://localhost:8000/v2/nodes/pvx9d/",
                     "meta":{
                        
                     }
                  }
               }
            },
            "registration_schema":{
               "data":{
                  "id":"5f52a449108c26016872e4ec",
                  "type":"registration-schemas"
               },
               "links":{
                  "related":{
                     "href":"http://localhost:8000/v2/schemas/registrations/5f52a449108c26016872e4ec/",
                     "meta":{
                        
                     }
                  }
               }
            },
            "requests":{
               "links":{
                  "related":{
                     "href":"http://localhost:8000/v2/registrations/6a9k8/requests/",
                     "meta":{
                        
                     }
                  }
               }
            },
            "review_actions":{
               "links":{
                  "related":{
                     "href":"http://localhost:8000/v2/registrations/6a9k8/actions/",
                     "meta":{
                        
                     }
                  }
               }
            },
            "root":{
               "data":{
                  "id":"6a9k8",
                  "type":"registrations"
               },
               "links":{
                  "related":{
                     "href":"http://localhost:8000/v2/registrations/6a9k8/",
                     "meta":{
                        
                     }
                  }
               }
            },
            "view_only_links":{
               "links":{
                  "related":{
                     "href":"http://localhost:8000/v2/registrations/6a9k8/view_only_links/",
                     "meta":{
                        
                     }
                  }
               }
            },
            "wikis":{
               "links":{
                  "related":{
                     "href":"http://localhost:8000/v2/registrations/6a9k8/wikis/",
                     "meta":{
                        
                     }
                  }
               }
            }
         },
         "type":"registrations"
      }
   ],
   "links":{
      "first":"None",
      "last":"None",
      "meta":{
         "per_page":10,
         "total":1
      },
      "next":"None",
      "prev":"None"
   },
   "meta":{
      "version":"2.0"
   }
}
}
```

** RegistrationProviderRequestList **

```json
{
   "data":[
      {
         "id":"5f60d1c963d3d300652f3936",
         "type":"registration-requests",
         "attributes":{
            "request_type":"withdrawal",
            "machine_state":"initial",
            "comment":"None",
            "created":"2020-09-15T14:38:01.081405",
            "modified":"2020-09-15T14:38:01.081449",
            "date_last_transitioned":"None"
         },
         "relationships":{
            "creator":{
               "links":{
                  "related":{
                     "href":"http://localhost:8000/v2/users/k4hv2/",
                     "meta":{
                        
                     }
                  }
               },
               "data":{
                  "id":"k4hv2",
                  "type":"users"
               }
            },
            "target":{
               "links":{
                  "related":{
                     "href":"http://localhost:8000/v2/registrations/r6x4z/",
                     "meta":{
                        
                     }
                  }
               },
               "data":{
                  "id":"r6x4z",
                  "type":"registrations"
               }
            }
         },
         "links":{
            "self":"http://localhost:8000/v2/requests/5f60d1c963d3d300652f3936/",
            "target":"http://localhost:8000/v2/registrations/r6x4z/"
         }
      }
   ],
   "links":{
      "first":"None",
      "last":"None",
      "prev":"None",
      "next":"None",
      "meta":{
         "total":1,
         "per_page":10
      }
   },
   "meta":{
      "version":"2.0"
   }
}
```

** RegistrationActionList **
```
{
   "data":[
      {
         "attributes":{
            "auto":false,
            "comment":"",
            "date_created":"2020-10-06T12:50:46.037505Z",
            "date_modified":"2020-10-06T12:50:46.037558Z",
            "from_state":"initial",
            "to_state":"pending",
            "trigger":"submit",
            "visible":true
         },
         "embeds":{
            "creator":{
               "data":{
                  "attributes":{
                     "active":true,
                     "date_registered":"2020-10-06T12:50:36.465708",
                     "education":[
                        
                     ],
                     "employment":[
                        
                     ],
                     "family_name":"Mercury1",
                     "full_name":"Freddie Mercury1",
                     "given_name":"Freddie",
                     "locale":"en_US",
                     "middle_names":"",
                     "social":{
                        
                     },
                     "suffix":"",
                     "timezone":"Etc/UTC"
                  },
                  "id":"5g4d3",
                  "links":{
                     "html":"http://localhost:5000/5g4d3/",
                     "profile_image":"https://secure.gravatar.com/avatar/bc6f4530022ee2424f4b73f7764d0733?d=identicon",
                     "self":"http://localhost:8000/v2/users/5g4d3/"
                  },
                  "relationships":{
                     "groups":{
                        "links":{
                           "related":{
                              "href":"http://localhost:80/v2/users/5g4d3/groups/",
                              "meta":{
                                 
                              }
                           }
                        }
                     },
                     "institutions":{
                        "links":{
                           "related":{
                              "href":"http://localhost:80/v2/users/5g4d3/institutions/",
                              "meta":{
                                 
                              }
                           },
                           "self":{
                              "href":"http://localhost:80/v2/users/5g4d3/relationships/institutions/",
                              "meta":{
                                 
                              }
                           }
                        }
                     },
                     "nodes":{
                        "links":{
                           "related":{
                              "href":"http://localhost:80/v2/users/5g4d3/nodes/",
                              "meta":{
                                 
                              }
                           }
                        }
                     },
                     "preprints":{
                        "links":{
                           "related":{
                              "href":"http://localhost:80/v2/users/5g4d3/preprints/",
                              "meta":{
                                 
                              }
                           }
                        }
                     },
                     "quickfiles":{
                        "links":{
                           "download":{
                              "href":"http://localhost:7777/v1/resources/snt8w5dmay/providers/osfstorage/?zip=",
                              "meta":{
                                 
                              }
                           },
                           "related":{
                              "href":"http://localhost:80/v2/users/5g4d3/quickfiles/",
                              "meta":{
                                 
                              }
                           },
                           "upload":{
                              "href":"http://localhost:7777/v1/resources/snt8w5dmay/providers/osfstorage/",
                              "meta":{
                                 
                              }
                           }
                        }
                     },
                     "registrations":{
                        "links":{
                           "related":{
                              "href":"http://localhost:80/v2/users/5g4d3/registrations/",
                              "meta":{
                                 
                              }
                           }
                        }
                     }
                  },
                  "type":"users"
               }
            }
         },
         "id":"5f7c6826d99cee00b89c372b",
         "links":{
            "self":"http://localhost:8000/v2/actions/5f7c6826d99cee00b89c372b/"
         },
         "relationships":{
            "creator":{
               "data":{
                  "id":"5g4d3",
                  "type":"users"
               },
               "links":{
                  "related":{
                     "href":"http://localhost:8000/v2/users/5g4d3/",
                     "meta":{
                        
                     }
                  }
               }
            },
            "target":{
               "data":{
                  "id":"fcp39",
                  "type":"registrations"
               },
               "links":{
                  "related":{
                     "href":"http://localhost:8000/v2/registrations/fcp39/",
                     "meta":{
                        
                     }
                  }
               }
            }
         },
         "type":"registration-actions"
      }
   ],
   "links":{
      "first":"None",
      "last":"None",
      "meta":{
         "per_page":10,
         "total":1
      },
      "next":"None",
      "prev":"None"
   },
   "meta":{
      "version":"2.0"
   }
}
```

** RegistrationRequestList **
```
{
   "data":[
      {
         "attributes":{
            "auto":false,
            "comment":"",
            "date_created":"2020-10-06T12:50:46.037505Z",
            "date_modified":"2020-10-06T12:50:46.037558Z",
            "from_state":"initial",
            "to_state":"pending",
            "trigger":"submit",
            "visible":true
         },
         "embeds":{
            "creator":{
               "data":{
                  "attributes":{
                     "active":true,
                     "date_registered":"2020-10-06T12:50:36.465708",
                     "education":[
                        
                     ],
                     "employment":[
                        
                     ],
                     "family_name":"Mercury1",
                     "full_name":"Freddie Mercury1",
                     "given_name":"Freddie",
                     "locale":"en_US",
                     "middle_names":"",
                     "social":{
                        
                     },
                     "suffix":"",
                     "timezone":"Etc/UTC"
                  },
                  "id":"5g4d3",
                  "links":{
                     "html":"http://localhost:5000/5g4d3/",
                     "profile_image":"https://secure.gravatar.com/avatar/bc6f4530022ee2424f4b73f7764d0733?d=identicon",
                     "self":"http://localhost:8000/v2/users/5g4d3/"
                  },
                  "relationships":{
                     "groups":{
                        "links":{
                           "related":{
                              "href":"http://localhost:80/v2/users/5g4d3/groups/",
                              "meta":{
                                 
                              }
                           }
                        }
                     },
                     "institutions":{
                        "links":{
                           "related":{
                              "href":"http://localhost:80/v2/users/5g4d3/institutions/",
                              "meta":{
                                 
                              }
                           },
                           "self":{
                              "href":"http://localhost:80/v2/users/5g4d3/relationships/institutions/",
                              "meta":{
                                 
                              }
                           }
                        }
                     },
                     "nodes":{
                        "links":{
                           "related":{
                              "href":"http://localhost:80/v2/users/5g4d3/nodes/",
                              "meta":{
                                 
                              }
                           }
                        }
                     },
                     "preprints":{
                        "links":{
                           "related":{
                              "href":"http://localhost:80/v2/users/5g4d3/preprints/",
                              "meta":{
                                 
                              }
                           }
                        }
                     },
                     "quickfiles":{
                        "links":{
                           "download":{
                              "href":"http://localhost:7777/v1/resources/snt8w5dmay/providers/osfstorage/?zip=",
                              "meta":{
                                 
                              }
                           },
                           "related":{
                              "href":"http://localhost:80/v2/users/5g4d3/quickfiles/",
                              "meta":{
                                 
                              }
                           },
                           "upload":{
                              "href":"http://localhost:7777/v1/resources/snt8w5dmay/providers/osfstorage/",
                              "meta":{
                                 
                              }
                           }
                        }
                     },
                     "registrations":{
                        "links":{
                           "related":{
                              "href":"http://localhost:80/v2/users/5g4d3/registrations/",
                              "meta":{
                                 
                              }
                           }
                        }
                     }
                  },
                  "type":"users"
               }
            }
         },
         "id":"5f7c6826d99cee00b89c372b",
         "links":{
            "self":"http://localhost:8000/v2/actions/5f7c6826d99cee00b89c372b/"
         },
         "relationships":{
            "creator":{
               "data":{
                  "id":"5g4d3",
                  "type":"users"
               },
               "links":{
                  "related":{
                     "href":"http://localhost:8000/v2/users/5g4d3/",
                     "meta":{
                        
                     }
                  }
               }
            },
            "target":{
               "data":{
                  "id":"fcp39",
                  "type":"registrations"
               },
               "links":{
                  "related":{
                     "href":"http://localhost:8000/v2/registrations/fcp39/",
                     "meta":{
                        
                     }
                  }
               }
            }
         },
         "type":"registration-actions"
      }
   ],
   "links":{
      "first":"None",
      "last":"None",
      "meta":{
         "per_page":10,
         "total":1
      },
      "next":"None",
      "prev":"None"
   },
   "meta":{
      "version":"2.0"
   }
}
```

## Side Effects

None that I know of.

## Ticket

https://openscience.atlassian.net/browse/ENG-2129

### CODE REVIEW

File: osf/migrations/0215_auto_20200901_1344.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+class Migration(migrations.Migration):
+
+    dependencies = [
+        ('osf', '0214_auto_20200701_1658'),
+    ]
+
+    operations = [
+        migrations.CreateModel(
+            name='RegistrationRequestAction',
```
Comment (brianjgeiger) [inline osf/migrations/0215_auto_20200901_1344.py:20]: Drive-by: I thought we didn't need the RequestActions. Is this going to change what we told front end, and is it going to affect the 'Integrate state machine' PR?

Comment (Johnetordoff) [reply osf/migrations/0215_auto_20200901_1344.py]: ug good catch this was should be disappeared because it's just duplicated because the auto-generated names were different. So when I pulled from develop this didn't go away. But It should be fixed now.

File: api/providers/views.py [hunk 2]
```diff
@@ -619,3 +637,108 @@ def get_default_queryset(self):
...
+        base_permissions.TokenHasScope,
+        MustBeModerator,
+    )
+    view_category = 'actions'
+    view_name = 'registration-provider-action-list'
+
+    required_read_scopes = [CoreScopes.ACTIONS_READ]
+    required_write_scopes = [CoreScopes.NULL]
+
+    serializer_class = NodeRequestActionSerializer
```
Comment (UdayVarkhedkar) [inline api/providers/views.py:726]: Will the write endpoints require the creation of a `RegistrationRequestActionSerializer`?

File: api/providers/views.py [hunk 3]
```diff
@@ -619,3 +637,108 @@ def get_default_queryset(self):
...
+        # used in perms class
+        return get_object_or_error(
+            RegistrationProvider,
+            self.kwargs['provider_id'],
+            self.request,
+            display_name='RegistrationProvider',
+        )
+
+    def get_default_queryset(self):
+        return NodeRequest.objects.filter(
```
Comment (UdayVarkhedkar) [inline api/providers/views.py:705]: Why should we use `NodeRequest` as the object here? Is it fine that the associated machine class for the object is the NodeRequestMachine?

**Review by UdayVarkhedkar — COMMENTED** [review]
Overall things look good but I have a few clarifying questions to help me better understand some of the decisions and how future tickets will build upon this work.

File: api/providers/views.py [hunk 4]
```diff
@@ -619,3 +637,108 @@ def get_default_queryset(self):
...
+    """
+    permission_classes = (
+        drf_permissions.IsAuthenticated,
+        base_permissions.TokenHasScope,
+        MustBeModerator,
+    )
+
+    ordering = ('-created')
+
+    serializer_class = DraftRegistrationSerializer
```
[the remark this answers is not in the document]
Comment (Johnetordoff) [reply api/providers/views.py:653]: yup, that's correct.

File: api/providers/views.py (hunk 2 above)
Comment (Johnetordoff) [reply api/providers/views.py:722]: Hmm... I don't think so, that should be figured in the integrate model methods ticket.

File: api/providers/views.py (hunk 3 above)
Comment (Johnetordoff) [reply api/providers/views.py:699]: This kind of just weird because NodeRequest has it's own machine_class, but it really shouldn't change or effect this too much. The idea here is that the front-end is going to be able to see all the request, but basically only needs the withdrawal ones, and it can just filter for them.

File: api/providers/views.py [hunk 5]
```diff
@@ -619,3 +637,108 @@ def get_default_queryset(self):
...
+
+    required_read_scopes = [CoreScopes.NODE_REGISTRATIONS_READ]
+    required_write_scopes = [CoreScopes.NULL]
+
+    view_category = 'registration-providers'
+    view_name = 'registrations-list'
+
+    def get_provider(self):
```
[the remark this answers is not in the document]
Comment (Johnetordoff) [reply api/providers/views.py]: Good idea! Done.

**Review by UdayVarkhedkar — APPROVED** [review]

**Review by brianjgeiger — CHANGES_REQUESTED** [review]
It would be useful for review, but especially for FE, to have a list of the endpoints that this enables with sample output and under what circumstances it would be used. That would make it easy for FE to set up mirage for their work, and it would enable them to make any additional requests for relationships or whatever else might make their lives easier.

File: api/providers/views.py [hunk 6]
```diff
@@ -611,11 +643,90 @@ class RegistrationProviderSchemaList(JSONAPIBaseView, generics.ListAPIView, List
...
+    required_write_scopes = [CoreScopes.NULL]
+
+    view_category = 'registration-providers'
+    view_name = 'registrations-list'
+
+    def get_default_queryset(self):
+        return DraftRegistration.objects.filter(
+            provider=self.get_provider(),
+            registered_node__registration_approval__state=RegistrationApproval.APPROVED,
+        )
```
Comment (brianjgeiger) [inline api/providers/views.py:677]: Shouldn't this actually be resolving to the Registration instead, though? I thought we had a discussion about that when we were deciding whether we could attach this to the AbstractNode or not, and that we'd have to take the performance hit and resolve these to the Registrations. 

I think it's going to confuse the ember app if we have it be the draft registration we serialize here, but actually being a registration.

File: api/providers/views.py [hunk 7]
```diff
@@ -611,11 +643,90 @@ class RegistrationProviderSchemaList(JSONAPIBaseView, generics.ListAPIView, List
...
+
+class RegistrationProviderRequestList(JSONAPIBaseView, generics.ListAPIView, ListFilterMixin, ProviderMixin):
+    provider_class = RegistrationProvider
+    permission_classes = (
+        drf_permissions.IsAuthenticated,
+        base_permissions.TokenHasScope,
+        MustBeModerator,
+    )
+    view_category = 'requests'
+    view_name = 'registration-provider-withdrawal-request-list'
```
Comment (brianjgeiger) [inline api/providers/views.py:692]: Should this be 'registration-provider-request-list'?

**Review by brianjgeiger — CHANGES_REQUESTED** [review]
One tiny thing, one bigger thing.

File: api/providers/views.py [hunk 8]
```diff
@@ -402,14 +436,15 @@ def get_renderer_context(self):
             # TODO don't duplicate the above
             auth = get_user_auth(self.request)
             auth_user = getattr(auth, 'user', None)
-            provider = get_object_or_error(PreprintProvider, self.kwargs['provider_id'], self.request, display_name='PreprintProvider')
+            provider = self.get_provider()
             if auth_user and auth_user.has_perm('view_submissions', provider):
                 context['meta'] = {
                     'reviews_state_counts': provider.get_reviewable_state_counts(),
                 }
```
Comment (brianjgeiger) [inline api/providers/views.py:443]: Is this available to registrations as well? The `meta.reviews_state_counts` was mentioned on the [list of things that FR had from preprints](https://www.notion.so/Registries-Moderation-Analysis-8b9a80f0fdb04da98a251112f4bf09be#0f55f0f90a61460ba591e5bd421291ce) that they'd likely need here.

**Review by brianjgeiger — CHANGES_REQUESTED** [review]
Found a potential item from the document that FE prepared for us. I think everything else looks complete.

File: api/draft_registrations/serializers.py [hunk 9]
```diff
@@ -35,10 +36,16 @@ class DraftRegistrationSerializer(DraftRegistrationLegacySerializer, Taxonomizab
...
+
+    filterable_fields = frozenset([
+        'machine_state',
+    ])
+
     category_choices = list(settings.NODE_CATEGORY_MAP.items())
     category_choices_string = ', '.join(["'{}'".format(choice[0]) for choice in category_choices])
 
     title = ser.CharField(required=False, allow_blank=True)
+    machine_state = ser.ChoiceField(read_only=True, required=False, choices=RegistrationStates.choices())
```
Comment (adlius) [inline api/draft_registrations/serializers.py:48]: `DraftRegistration` probably doesn't need this.

File: api/providers/views.py [hunk 10]
```diff
@@ -611,11 +643,90 @@ class RegistrationProviderSchemaList(JSONAPIBaseView, generics.ListAPIView, List
...
-
     def get_default_queryset(self):
         return self.get_provider().schemas.get_latest_versions(request=self.request, invisible=True).filter(active=True)
 
     def get_queryset(self):
         return self.get_queryset_from_request()
+
+
+class RegistrationProviderRegistrationList(JSONAPIBaseView, generics.ListAPIView, ListFilterMixin, ProviderMixin):
+    """The documentation for this endpoint can be found [here](https://developer.osf.io/#operation/preprint_providers_preprints_list).
```
Comment (adlius) [inline api/providers/views.py:654]: Copy-pasta

**Comment by adlius** [comment]
Might make sense to add `requests` and `actions` relationship to the registration serializer.

File: api/providers/views.py [hunk 11]
```diff
@@ -611,11 +643,100 @@ class RegistrationProviderSchemaList(JSONAPIBaseView, generics.ListAPIView, List
...
+        base_permissions.TokenHasScope,
+        MustBeModerator,
+    )
+    view_category = 'actions'
+    view_name = 'registration-provider-action-list'
+
+    required_read_scopes = [CoreScopes.ACTIONS_READ]
+    required_write_scopes = [CoreScopes.NULL]
+
+    serializer_class = NodeRequestActionSerializer
```
Comment (adlius) [inline api/providers/views.py:733]: If the `serializer_class` is `NodeRequestActionSerializer`, that means the `type` for this resource is `node-request-actions`. Does that mean every such action is related to a request?

File: osf/models/provider.py [hunk 12]
```diff
@@ -198,6 +199,9 @@ def absolute_api_v2_url(self):
 
 
 class RegistrationProvider(AbstractProvider):
+    REVIEWABLE_RELATION_NAME = 'registrations'
```
Comment (jwalz) [inline osf/models/provider.py:202]: drive by: what would the consequence of just making this 'draft_registrations' be? That would eliminate the need for the MODERATION_MACHINE_STATE_RELATION, correct?

Comment (jwalz) [reply osf/models/provider.py:202]: Am I correct in understanding that this is helping us annotate the registration itself with the state?

Comment (Johnetordoff) [reply osf/models/provider.py]: 1. Yep! good catch
2. Yep, the only real relevancy of that was in `get_reviewable_state_counts` but as you point that function can just as easily grab a DraftRegistration as a Registration, it just didn't occur to me initially because the serializer is serializing registrations.

**Review by brianjgeiger — APPROVED** [review]
